### PR TITLE
Implimentation of Dual-Slope signal scaling for the Inner Pixel

### PR DESCRIPTION
**Pull request by suchandradutta**
Updating Phase2TrackerDigitizerAlgorithm with the option of dual slope scaling of signal to be used for the inner pixel detector. A new monitoring histogram added in one of the analyzers in the validation suite (Phase2TrackerMonitorDigi).

The update was present in the Phase2 tracker simulation meeting
https://indico.cern.ch/event/688845/contributions/3026531/attachments/1660515/2660028/DualSlopeCorrection_May30.pdf

NB: There will be a similar PR for the Phase2 clusterizer both these PRs should be included together

@emiglior @avartak @boudoul

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by kpedro88** [comment]
```
20-Aug-2018 17:56:03 CEST  Initiating request to open file file:step1.root
20-Aug-2018 17:56:04 CEST  Successfully opened file file:step1.root
20-Aug-2018 17:56:06 CEST  Closed file file:step1.root
----- Begin Fatal Exception 20-Aug-2018 17:56:06 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=MixingModule label='mix'
Exception Message:
MissingParameter: Parameter 'ReadoutMode' not found.
----- End Fatal Exception -------------------------------------------------
```

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 1]
```diff
@@ -989,3 +993,32 @@ void Phase2TrackerDigitizerAlgorithm::digitize(const Phase2TrackerGeomDetUnit* p
...
+  int signal_in_adc;
+  int temp_signal;
+  if (thePhase2ReadoutMode == 0) signal_in_adc = theAdcFullScale;
+  else { 
+    
+    if (thePhase2ReadoutMode == -1) {
+      temp_signal = std::floor(signal_in_elec / theElectronPerADC) + 1;
+    } else {
+      // calculate the kink point and the slope
+      const int dualslope_param = std::min(abs(thePhase2ReadoutMode), 10);
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: make a named constant for magic number 10

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.h [hunk 2]
```diff
@@ -116,7 +116,7 @@ class Phase2TrackerDigitizerAlgorithm  {
   const float ClusterWidth;       // Gaussian charge cutoff width in sigma units
   
   //-- make_digis 
-  const bool doDigitalReadout;        //  Flag to decide analog or digital readout
+  const int   thePhase2ReadoutMode;  //  Flag to decide readout mode (digital/Analong dual slope etc.)
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.h]: analong -> analog

File: SimTracker/SiPhase2Digitizer/test/Phase2TrackerMonitorDigi.cc [hunk 3]
```diff
@@ -149,14 +150,19 @@ void Phase2TrackerMonitorDigi::fillITPixelDigiHistos(const edm::Handle<edm::DetS
...
       } else {
 	if (abs(row - row_last) == 1 && col == col_last) {
 	  position += row+1;
 	  width++;
+	  charges.push_back(adc);
 	} else {
           position /= width;  
           local_mes.ClusterWidth->Fill(width);
           local_mes.ClusterPosition->Fill(position);
+	  for (std::vector<int>::iterator ic = charges.begin(); ic != charges.end(); ic++) local_mes.ChargeOfDigisVsWidth->Fill((*ic), width);
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/test/Phase2TrackerMonitorDigi.cc]: use range-based loop

**Comment by kpedro88** [comment]
please test

**Comment by civanch** [comment]
please test

**Comment by kpedro88** [comment]
please address comment https://github.com/cms-sw/cmssw/pull/24325#discussion_r211409033

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by kpedro88** [comment]
+upgrade
minor changes in Phase2 workflows consistent with modification of pixel digis

**Comment by kpedro88** [comment]
@suchandradutta @avartak @emiglior @boudoul should we wait to merge this one until #24329 is ready?

**Comment by kpedro88** [comment]
But this PR does not 100% reproduce the previous implementation, since a number of (small) differences were observed in the comparisons. Is that understood?

**Comment by emiglior** [comment]
please test

**Comment by emiglior** [comment]
Screenshots of a dEdx distribution in the previous version of the PR [1] and in the new version of the PR [2].  In the new PR the differences wrt baseline are much reduced a part from small (expected) differences in the tail (we have slightly modified the max value of one parameter). 

[1] workflow20034-dedxdqmharm2po-originalpr
[2] workflow20034-dedxdqmharm2po-newpr

**Comment by kpedro88** [comment]
+upgrade

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**